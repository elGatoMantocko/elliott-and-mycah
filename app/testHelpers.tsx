import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { match } from 'css-mediaquery';
import React from 'react';
import { MemoryRouter, Route, Routes, useLocation } from 'react-router';

type TestLocations = 'route-location';

/**
 * The element testid used to show the location.
 */
export const TestRouteLocation: TestLocations = 'route-location';

interface TestLocationProps {
  testId?: string;
}

/**
 * Component that exposes the location path and search to the dom
 * @param root0 props
 * @param root0.testId test id passed to the inner location element
 * @returns element
 */
const TestLocation = ({ testId }: TestLocationProps) => {
  const location = useLocation();
  return (
    <div data-testid={testId ?? TestRouteLocation}>{`${location.pathname}${location.search}`}</div>
  );
};

interface RouterRenderOptions extends RenderOptions {
  locationTestId?: string;
}

/**
 * Use to render UI with a react-router with the ability to search for a `route-location` test id.
 * @example
 * ```typescriptreact
 * const el = renderWithRouter(<MyComponent />);
 * const locationElementData = await el.findByTestId(TestRouteLocation);
 * ```
 * @param ui to render in the test
 * @param options passed to testing-library
 * @param options.locationTestId passed to the `<TestLocation />` element
 * @returns render result
 */
export const renderWithRouter = (
  ui: React.ReactElement,
  { locationTestId, ...renderOptions }: RouterRenderOptions = {},
): RenderResult => {
  return render(
    <MemoryRouter>
      {ui}
      <Routes>
        <Route path="*" element={<TestLocation testId={locationTestId} />} />
      </Routes>
    </MemoryRouter>,
    renderOptions,
  );
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = () => {};

/**
 * Function to create a match media for mocking media queries.
 * @param width to mock in the jsdom
 * @returns matchMedia options
 */
export function createMatchMedia(width: number): (query: string) => MediaQueryList {
  return (query: string) =>
    ({
      matches: match(query, { width }),
      addListener: noop,
      removeListener: noop,
    }) as unknown as MediaQueryList;
}
