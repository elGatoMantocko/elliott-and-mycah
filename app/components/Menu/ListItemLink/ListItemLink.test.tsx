import { act, render, screen } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router';

import { renderWithRouter } from '../../../testHelpers';
import { ListItemLink } from '.';

it('should render a link', () => {
  const { asFragment } = render(<ListItemLink to="/test">Test</ListItemLink>, {
    wrapper: MemoryRouter,
  });
  expect(asFragment()).toMatchSnapshot();
});

it('should render a disabled link', async () => {
  renderWithRouter(
    <ListItemLink data-testid="disabled-link" disabled to="/test">
      Test
    </ListItemLink>,
    { locationTestId: 'disabled-location' },
  );

  const link = screen.getByTestId('disabled-link');
  expect(link).toMatchSnapshot('disabled-link');
  expect(link.className.split(' ')).toContain('Mui-disabled');

  const routeEl = screen.getByTestId('disabled-location');
  expect(routeEl.textContent).toEqual('/');
});

it('should navigate to href when clicked', async () => {
  const el = renderWithRouter(
    <ListItemLink data-testid="navigate-with-href" to="/test-path" onClick={undefined}>
      Test
    </ListItemLink>,
    { locationTestId: 'navigate-with-href-location' },
  );

  await act(async () => {
    const link = el.getByTestId('navigate-with-href');
    link.click();
  });

  // get the location element and assert it matches the expected path
  const routeEl = el.getByTestId('navigate-with-href-location');
  expect(routeEl.textContent).toEqual('/test-path');

  expect(el.getByTestId('navigate-with-href')).toMatchSnapshot('navigate-with-href');
});

it('should call the onClick handler', async () => {
  const onClickMock = jest.fn();
  const el = renderWithRouter(
    <ListItemLink to="/test" data-testid="navigate-with-onClick" onClick={onClickMock}>
      Test
    </ListItemLink>,
    { locationTestId: 'call-onclick-location' },
  );

  await act(async () => {
    const link = el.getByTestId('navigate-with-onClick');
    link.click();
  });

  expect(onClickMock).toHaveBeenCalled();

  // get the location element and assert it matches the expected path
  const routeEl = el.getByTestId('call-onclick-location');
  expect(routeEl.textContent).toEqual('/test');

  expect(el.getByTestId('navigate-with-onClick')).toMatchSnapshot('navigate-with-onClick');
});
