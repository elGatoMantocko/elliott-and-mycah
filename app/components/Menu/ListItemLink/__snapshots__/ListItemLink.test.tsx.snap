// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should call the onClick handler > navigate-with-onClick 1`] = `
<a
  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
  data-testid="navigate-with-onClick"
  href="/test"
  tabindex="0"
>
  Test
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</a>
`;

exports[`should navigate to href when clicked > navigate-with-href 1`] = `
<a
  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
  data-testid="navigate-with-href"
  href="/test-path"
  tabindex="0"
>
  Test
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</a>
`;

exports[`should render a disabled link > disabled-link 1`] = `
<a
  aria-disabled="true"
  class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-disabled Mui-disabled MuiListItemButton-root MuiListItemButton-gutters Mui-disabled css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
  data-testid="disabled-link"
  href="/test"
  tabindex="-1"
>
  Test
</a>
`;

exports[`should render a link 1`] = `
<DocumentFragment>
  <a
    class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
    href="/test"
    tabindex="0"
  >
    Test
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
</DocumentFragment>
`;
